Car entries: possibility to have the same number of entries per car type

Hi, first, thank you for the work you did !

When creating a new custom race, and adding let's say 24 entries and I have 8 cars available, if on each entry, I set the car to "any available car", I don't have the same number of cars per car type when starting the race. For this example, I would have 3 cars for each car type (3x8=24). But for the moment I have a random number (with at least 1 car for each type of car).

Maybe this could be an option to equally select random cars ...

I hope you understand what I mean. Maybe I could help, I don't know the go langage but I can try to learn ...

Thank you very much !

Hi,

I am trying to recreate the makefiles with a run.bat for Windows. It's seems easy because the makefiles are only command lines. I will try to do a pull request with the run.bat files so that people can build the project on windows.

I have an error though, when trying to do the run command which compiles the project:
go build -ldflags "-s -w -X github.com/JustaPenguin/assetto-server-manager.BuildVersion=%VERSION% %LDFLAGS%"

# github.com/JustaPenguin/assetto-server-manager/internal/changelog
..\..\internal\changelog\changelog.go:13:20: undefined: FSByte

I have opened the source file, as I don't really know Go, I am not sure how to correct this. Is it a Linux dependency or something inside the code ? 

Thank you

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,7 @@ Added:
 * Track images now show the track map in the bottom right corner.
 * New lua plugin file (race-control.lua) containing onChat function which is called whenever a chat message is sent in-game, from the live timings page or from a lua script
 * (Multiserver) Events across all servers will now be shown on the Calendar, events on other servers will be marked as so. 
+* Events using "Any Available Car" now split the car types evenly between the number of entrants.
 
 Fixed:
 

--- a/cmd/server-manager/typescript/src/Font.ts
+++ b/cmd/server-manager/typescript/src/Font.ts
@@ -50,6 +50,7 @@ import {faShieldAlt} from "@fortawesome/free-solid-svg-icons/faShieldAlt"
 import {faVideo} from "@fortawesome/free-solid-svg-icons/faVideo";
 import {faScroll} from "@fortawesome/free-solid-svg-icons/faScroll"
 import {faRedoAlt} from "@fortawesome/free-solid-svg-icons/faRedoAlt";
+import {faClock} from "@fortawesome/free-solid-svg-icons/faClock";
 
 library.add(
     faUser,
@@ -101,6 +102,7 @@ library.add(
     faVideo,
     faScroll,
     faRedoAlt,
+    faClock,
 );
 
 dom.watch();

--- a/cmd/server-manager/views/pages/custom-race/index.html
+++ b/cmd/server-manager/views/pages/custom-race/index.html
@@ -175,10 +175,18 @@
 
     <nav>
         <div class="nav nav-tabs" id="nav-tab" role="tablist">
-            <a class="nav-item nav-link active" id="nav-recent-tab" data-toggle="tab" href="#nav-recent" role="tab" aria-controls="nav-recent" aria-selected="false">Recent</a>
-            <a class="nav-item nav-link" id="nav-starred-tab" data-toggle="tab" href="#nav-starred" role="tab" aria-controls="nav-starred" aria-selected="true">Starred</a>
-            <a class="nav-item nav-link" id="nav-scheduled-tab" data-toggle="tab" href="#nav-scheduled" role="tab" aria-controls="nav-scheduled" aria-selected="false">Scheduled</a>
-            <a class="nav-item nav-link" id="nav-autoloop-tab" data-toggle="tab" href="#nav-autoloop" role="tab" aria-controls="nav-autoloop" aria-selected="false">Auto Loop</a>
+            <a class="nav-item nav-link active" id="nav-recent-tab" data-toggle="tab" href="#nav-recent" role="tab" aria-controls="nav-recent" aria-selected="false">
+                <i class="fas fa-clock mr-1"></i> Recent
+            </a>
+            <a class="nav-item nav-link" id="nav-starred-tab" data-toggle="tab" href="#nav-starred" role="tab" aria-controls="nav-starred" aria-selected="true">
+                <i class="fas fa-star mr-1"></i> Starred
+            </a>
+            <a class="nav-item nav-link" id="nav-scheduled-tab" data-toggle="tab" href="#nav-scheduled" role="tab" aria-controls="nav-scheduled" aria-selected="false">
+                <i class="fas fa-calendar mr-1"></i> Scheduled
+            </a>
+            <a class="nav-item nav-link" id="nav-autoloop-tab" data-toggle="tab" href="#nav-autoloop" role="tab" aria-controls="nav-autoloop" aria-selected="false">
+                <i class="fas fa-play-circle mr-1"></i> Auto Loop
+            </a>
         </div>
     </nav>
     <div class="tab-content" id="nav-tabContent">

--- a/race_manager.go
+++ b/race_manager.go
@@ -249,22 +249,15 @@ func (rm *RaceManager) applyConfigAndStart(event RaceEvent) error {
 		return err
 	}
 
-	// any available car should make sure you have one of each before randomising (#678)
-	for _, car := range finalCars {
-		for _, entrant := range entryList {
-			if entrant.Model == AnyCarModel {
-				entrant.Model = car
-				entrant.Skin = rm.carManager.RandomSkin(entrant.Model)
-				break
-			}
-		}
-	}
+	numEntrantsWithAnyCar := 0
 
 	for _, entrant := range entryList {
 		if entrant.Model == AnyCarModel {
 			// cars with 'any car model' become random in the entry list.
-			entrant.Model = finalCars[rand.Intn(len(finalCars))]
+			entrant.Model = finalCars[numEntrantsWithAnyCar%len(finalCars)]
 			entrant.Skin = rm.carManager.RandomSkin(entrant.Model)
+
+			numEntrantsWithAnyCar++
 		}
 	}
 

